core/src/main/java/org/testcontainers/containers/GenericContainer.java

```diff
@@ -718,21 +720,32 @@ protected Integer getLivenessCheckPort() {
 
     private void applyConfiguration(CreateContainerCmd createCommand) {
         HostConfig hostConfig = buildHostConfig();
-        createCommand.withHostConfig(hostConfig);
 
-        // Set up exposed ports (where there are no host port bindings defined)
-        ExposedPort[] portArray = exposedPorts.stream()
-                .map(ExposedPort::new)
-                .toArray(ExposedPort[]::new);
+        // PortBindings must contain:
+        //  * all exposed ports with a randomized host port (equivalent to -p CONTAINER_PORT)
+        //  * all exposed ports with a fixed host port (equivalent to -p HOST_PORT:CONTAINER_PORT)
+        List<PortBinding> allPortBindings = new ArrayList<>();
+        // First collect all the randomized host ports from our 'exposedPorts' field
+        exposedPorts.stream()
+            .map(ExposedPort::new)
+            .map(p -> new PortBinding(Ports.Binding.empty(), p))
+            .forEachOrdered(allPortBindings::add);
+        // Next collect all the fixed host ports from our 'portBindings' field
+        portBindings.stream()
+            .map(PortBinding::parse)
+            .forEachOrdered(allPortBindings::add);
+
+        hostConfig.withPortBindings(allPortBindings);
+
+        // Next, ExposedPorts must be set up to publish all of the above ports, randomized and fixed.
+        // Collect all of the exposed ports for publication
+        final List<ExposedPort> exposedPorts = allPortBindings.stream()
+            .map(PortBinding::getExposedPort)
+            .distinct()
+            .collect(Collectors.toList());
```

Comment:
This is the thing that is substantially different from the previous implementation; changes above are mainly to support this change and to make it clearer what we're doing.

Of course, `createCommand.withPublishAllPorts(true);` on line 801 is also nuked.